### PR TITLE
Object store validation

### PR DESCRIPTION
**Pull request by replay**
We have a customer with a problem which we suspected to happen because their on-prem S3 alternative was returning incorrect results. This tool has proven to be useful in debugging that issue, so I figured we might want to merge it because it could be useful when debugging future issues again.

### CODE REVIEW

**Review by simonswine — COMMENTED** [review]
Seems good to me, I guess in general it would be nice getting a bit more progress update. 

Was not entirely sure if it was stuck or still running even with debug flags on.

Also another thought: It would be nice to use `go test` primitives, but they might not integrate that nicely with a golang main.

File: pkg/commands/bucket_validation.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	b.objectNames = make(map[string]string, b.objectCount)
+	for objectIdx := 0; objectIdx < b.objectCount; objectIdx++ {
+		b.objectNames[fmt.Sprintf("%05X/", objectIdx)] = "testfile"
+	}
+}
+
+func (b *BucketValidationCommand) createTestData(ctx context.Context) error {
+	for dirName, objectName := range b.objectNames {
+		objectPath := dirName + objectName
+		err := b.bucketClient.Upload(ctx, objectPath, bytes.NewReader([]byte(b.objectContent)))
```
[the remark this answers is not in the document]
Comment (replay) [reply pkg/commands/bucket_validation.go:107]: interesting!

**Comment by replay** [comment]
>  I guess in general it would be nice getting a bit more progress update.

Good idea, will add that.

> Also another thought: It would be nice to use go test primitives

That's a nice idea, for now I'll not implement it because I want to get this out to the customer without spending much more time, but it would certainly be a "nice to have" for the future

**Comment by replay** [comment]
I added some status updates to it, and I also removed the `--verbose` parameter because using the `strings.Reader` the bucket client isn't spamming the terminal with those warnings anymore. The output now looks like this:

```
replay@nb-ubuntu:~/go/src/github.com/grafana/cortex-tools$ ./cmd/cortextool/cortextool bucket-validation validate  --s3.bucket-name=testbucket --s3.access-key-id=mykey --s3.secret-access-key=mysecret --s3.endpoint=localhost:9000 --s3.signature-version=v2 --s3.insecure --object-count=200 --test-runs=2 --report-every=50
level=info phase="creating test objects" completed=0 total=200
level=info phase="creating test objects" completed=50 total=200
level=info phase="creating test objects" completed=100 total=200
level=info phase="creating test objects" completed=150 total=200
level=info phase="creating test objects" completed=200 total=200
level=info phase="listing test objects"
level=info phase="validating test objects" completed=0 total=200
level=info phase="validating test objects" completed=50 total=200
level=info phase="validating test objects" completed=100 total=200
level=info phase="validating test objects" completed=150 total=200
level=info phase="validating test objects" completed=200 total=200
level=info phase="deleting test objects" completed=0 total=200
level=info phase="deleting test objects" completed=50 total=200
level=info phase="deleting test objects" completed=100 total=200
level=info phase="deleting test objects" completed=150 total=200
level=info phase="deleting test objects" completed=200 total=200
level=info testrun_successful=1
level=info phase="creating test objects" completed=0 total=200
level=info phase="creating test objects" completed=50 total=200
level=info phase="creating test objects" completed=100 total=200
level=info phase="creating test objects" completed=150 total=200
level=info phase="creating test objects" completed=200 total=200
level=info phase="listing test objects"
level=info phase="validating test objects" completed=0 total=200
level=info phase="validating test objects" completed=50 total=200
level=info phase="validating test objects" completed=100 total=200
level=info phase="validating test objects" completed=150 total=200
level=info phase="validating test objects" completed=200 total=200
level=info phase="deleting test objects" completed=0 total=200
level=info phase="deleting test objects" completed=50 total=200
level=info phase="deleting test objects" completed=100 total=200
level=info phase="deleting test objects" completed=150 total=200
level=info phase="deleting test objects" completed=200 total=200
level=info testrun_successful=2
```

**Comment by replay** [comment]
FYI I rebased onto `main`

**Comment by replay** [comment]
I cleaned it up a bit and I think it should be ready for review now.

Unfortunately the config parsing is a big ugly because `cortex-tool` uses `kingpin` for config parsing, while in the Cortex uses the normal `flag` package. [So I used a kingpin argument which takes the config string that's then passed to the flagset configuring the Cortex config struct](https://github.com/grafana/cortex-tools/blob/a7af75ce9b28b7897e45e9273e09955dfd2b4464/pkg/commands/bucket_validation.go#L99), I feel dirty for doing that but I couldn't see a better way.

File: pkg/commands/bucket_validation.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+func (b *BucketValidationCommand) validateTestObjects(ctx context.Context) error {
+	foundDirs := make(map[string]struct{}, b.objectCount)
+
+	level.Info(b.logger).Log("phase", "listing test objects")
+
+	err := b.bucketClient.Iter(ctx, b.prefix, func(dirName string) error {
+		foundDirs[dirName] = struct{}{}
+		return nil
+	})
+	if err != nil {
```
Comment (simonswine) [inline pkg/commands/bucket_validation.go:225]: Do you think it makes sense to compare the count of found test objects here?

Or at least add a output how many objects have been found. I just wondered what the log message is trying to tell only looking at the output.

**Review by simonswine — APPROVED** [review]
LGTM, small non blocking nit comment.

I like how the config is now quite aligned to cortex

File: pkg/commands/bucket_validation.go (hunk 2 above)
Comment (replay) [reply pkg/commands/bucket_validation.go:225]: ic, that's a good idea, I'll add that. thx

Comment (replay) [reply pkg/commands/bucket_validation.go:225]: I added the object count validation in two places (42bcb98882f873a80dec6c94fd930fecdc2b220b):
1) when we do the initial list in `validateTestObjects()`
2) right after each delete in `deleteTestObjects()`

**Pull request merged**